tools/test-runner/src/main.rs

```diff
@@ -0,0 +1,210 @@
+#[derive(Debug)]
+struct Test {
+    target: String,
+    command: process::Command,
+}
+
+fn run_all(specs: Vec<TestSpec>, threads: u32) -> Result<i32> {
+    // try to validate all test specs before running anything
+    let specs: Vec<Result<Test>> = specs.into_iter().map(validate).collect();
+    let mut tests: Vec<Test> = Vec::with_capacity(specs.len());
+
+    // print all errors and exit if there were any, while collecting valid tests
+    let mut quit = false;
+    for spec in specs {
+        match spec {
+            Err(e) => {
+                eprintln!("{}", e);
+                quit = true;
+            }
+            Ok(t) => {
+                tests.push(t);
+            }
+        };
+    }
+    if quit {
+        bail!("Found one or more invalid test specifications, see above");
+    }
+
+    // at this point, we should be able to safely run tests, even if they fail
+    let total = tests.len();
+    println!("Running {} test targets...", total);
+    let mut passed = 0;
+    let mut errors: Vec<String> = Vec::new();
+    for test in tests.iter_mut() {
+        let target = &test.target;
+        let mut child = test.command.spawn()?;
+        let status = child.wait()?;
+        if status.success() {
+            println!("[PASS] {}", target);
+            passed += 1;
+        } else {
+            println!("[FAIL] {} ({})", target, status);
+            let mut message = format!("\nTarget {} failed:\n", target);
+            let stderr = child.stderr.take().unwrap();
+            for line in BufReader::new(stderr).lines() {
+                let line = format!("    {}\n", line.unwrap());
+                message.push_str(&line);
+            }
+            errors.push(message);
+        }
+    }
+
+    // if there were any errors, we let the user know what they were
+    for error in errors {
+        eprintln!("{}", error);
+    }
+
+    // put a summary in the output, as well in this runner's return code
+    let percent = 100.0 * passed as f32 / total as f32;
+    println!(
+        "  Summary: {:.2}% test targets passed ({} out of {})",
+        percent, passed, total
+    );
+    let fail_count = (total - passed) as i32;
+    Ok(fail_count)
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "snake_case")]
+enum TestType {
+    Pyunit,
+}
+
+// refer to https://buck.build/files-and-dirs/buckconfig.html#test.external_runner
+// serde_json will expect AT LEAST these fields to be provided, ignoring the rest
+#[derive(Deserialize, Debug)]
+struct TestSpec {
+    /// The buck target URI of the test rule.
+    target: String,
+
+    /// The type of the test.
+    #[serde(rename(deserialize = "type"))]
+    type_: TestType,
+
+    /// Command line that should be used to run the test.
+    command: Vec<String>,
+
+    /// Environment variables to be defined when running the test.
+    env: HashMap<String, String>,
+
+    /// Labels that are defined on the test rule. NOTE: not used
+    labels: Vec<String>,
```

Comment:
Just leaving a note here - this is a good place if we ever need a way to control the test runner behavior from the unitttest target itself - it's a common pattern we use internally but I don't see a need for it in oss yet